src/utils/predefinedSelections.ts

```diff
@@ -0,0 +1,376 @@
+        return `${yearsAgo} year${yearsAgo === 1 ? "" : "s"} ago`;
+      }
+      const monthsAgo = differenceInMonths(now, date);
+      if (monthsAgo > 0) {
+        return `${monthsAgo} month${monthsAgo === 1 ? "" : "s"} ago`;
+      }
+      const weeksAgo = differenceInWeeks(now, date);
+      if (weeksAgo > 0) {
+        return `${weeksAgo} week${weeksAgo === 1 ? "" : "s"} ago`;
+      }
+      const daysAgo = differenceInDays(now, date);
+      if (daysAgo > 0) {
+        return `${daysAgo} day${daysAgo === 1 ? "" : "s"} ago`;
+      }
+      const hoursAgo = differenceInHours(now, date);
+      if (hoursAgo > 0) {
+        return `${hoursAgo} hour${hoursAgo === 1 ? "" : "s"} ago`;
+      }
+      const minutesAgo = differenceInMinutes(now, date);
+      if (minutesAgo > 0) {
+        return `${minutesAgo} minute${minutesAgo === 1 ? "" : "s"} ago`;
+      }
+      const secondsAgo = differenceInSeconds(now, date);
+      return `${secondsAgo} second${secondsAgo === 1 ? "" : "s"} ago`;
+    }
+    if (/time/i.test(arg)) {
+      return `${date.getHours().toString().padStart(2, "0")}:${date
+        .getMinutes()
+        .toString()
+        .padStart(2, "0")}`;
+    }
+    return date;
+  };
+  return {
+    "": date,
+    "-display": parseArg(),
+  };
+};
+
+const getBlockAttribute = (key: string, r: PullBlock) => {
+  const block = window.roamAlphaAPI.data.fast.q(
+    `[:find (pull ?b [:block/string :block/uid]) :where [?a :node/title "${normalizePageTitle(
+      key
+    )}"] [?p :block/uid "${
+      r[":block/uid"]
+    }"] [?b :block/refs ?a] [?b :block/parents ?p]]`
+  )?.[0]?.[0] as PullBlock;
+  return {
+    "": (block?.[":block/string"] || "").slice(key.length + 2).trim(),
+    "-uid": block?.[":block/uid"] || "",
+  };
+};
+
+const isVariableExposed = (
+  clauses: DatalogClause[],
+  variable: string
+): boolean =>
+  clauses.some((c) => {
+    switch (c.type) {
+      case "data-pattern":
+      case "fn-expr":
+      case "pred-expr":
+      case "rule-expr":
+        return c.arguments.some((a) => a.value === variable);
+      case "not-clause":
+      case "or-clause":
+      case "and-clause":
+        return isVariableExposed(c.clauses, variable);
+      case "not-join-clause":
+      case "or-join-clause":
+        return c.variables.some((v) => v.value === variable);
+      default:
+        return false;
+    }
+  });
+
+export type PredefinedSelection = {
+  test: RegExp;
+  pull: (a: {
+    returnNode: string;
+    match: RegExpExecArray | null;
+    where: DatalogClause[];
+  }) => string;
+  mapper: (
+    r: PullBlock,
+    key: string,
+    result: QueryResult
+  ) =>
+    | QueryResult[string]
+    | Record<string, QueryResult[string]>
+    | Promise<QueryResult[string] | Record<string, QueryResult[string]>>;
+  update?: (s: {
+    uid: string;
+    value: string;
+    // TODO - we wouldn't need this if `NODE_TEST` was separated out and not so overloaded.
+    selection: string;
+  }) => Promise<void>;
```

Comment:
Only notable changes in this file are searches for `update`.

In order to deterministically update query results, each selection will eventually need to implement an `update` method